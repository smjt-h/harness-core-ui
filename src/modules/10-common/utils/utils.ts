/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName } from 'framework/types/ModuleName'

interface SetPageNumberProps {
  setPage: (value: React.SetStateAction<number>) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export const isCommunityPlan = (): boolean => false

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined
