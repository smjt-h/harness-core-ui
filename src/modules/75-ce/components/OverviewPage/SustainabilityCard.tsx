/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import cx from 'classnames'
import { Color, FontVariation } from '@harness/design-system'
import { Container, Icon, Layout, Text } from '@harness/uicore'
import { String } from 'framework/strings'
import grayLeafImg from '@ce/common/images/gray-leaf.svg'
import css from './OverviewPage.module.scss'

interface SustainabilityCardProps {
  title: string
  firstColValue: string
  firstColText?: string
  secondColValue: string
  secondColText?: string
  className?: string
  fetching?: boolean
}

const SustainabilityCard: React.FC<SustainabilityCardProps> = props => {
  const { fetching, title, firstColValue, firstColText, secondColValue, secondColText, className } = props
  return (
    <div className={cx(css.sustainabilityCard, className)}>
      {fetching ? (
        <Layout.Horizontal flex={{ alignItems: 'center', justifyContent: 'center' }}>
          <Icon name="spinner" color="blue500" size={30} />
        </Layout.Horizontal>
      ) : (
        <>
          <Layout.Horizontal flex={{ alignItems: 'center', justifyContent: 'flex-start' }} spacing="small">
            <img src={grayLeafImg} width={16} />
            <Text font={{ variation: FontVariation.H6 }} color={Color.GREY_600}>
              {title}
            </Text>
          </Layout.Horizontal>
          <Layout.Horizontal spacing={'huge'} margin={{ top: 'var(--spacing-medium)' }}>
            <Container>
              <Text font={{ variation: FontVariation.H4 }} color={Color.GREY_800}>
                <String stringID="ce.common.emissionUnitHTML" vars={{ value: firstColValue }} useRichText />
              </Text>
              <Text font={{ variation: FontVariation.TINY }} color={Color.GREY_500}>
                {firstColText}
              </Text>
            </Container>
            <Container>
              <Text font={{ variation: FontVariation.H4 }} color={Color.GREY_800}>
                <String
                  stringID="ce.common.emissionUnitHTML"
                  vars={{
                    value: secondColValue
                  }}
                  useRichText
                />
              </Text>
              <Text font={{ variation: FontVariation.TINY }} color={Color.GREY_500}>
                {secondColText}
              </Text>
            </Container>
          </Layout.Horizontal>
        </>
      )}
    </div>
  )
}

export default SustainabilityCard
