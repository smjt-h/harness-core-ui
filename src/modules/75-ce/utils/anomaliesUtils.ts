/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { CloudProvider } from '@ce/types'
import type { QlceViewFilterInput } from 'services/ce/services'
import { DEFAULT_GROUP_BY } from './perspectiveUtils'

const gcpEntities = ['gcpProduct', 'gcpProjectId', 'gcpSKUDescription']
const awsEntities = ['awsUsageAccountId', 'awsServiceCode', 'awsInstancetype', 'awsUsageType']
const azureEntities: string[] = []
const clusterEntities = ['clusterName', 'clusterId', 'namespace', 'workloadName', 'workloadType']

const cloudProviderToEntityMapping = {
  [CloudProvider.GCP]: gcpEntities,
  [CloudProvider.AWS]: awsEntities,
  [CloudProvider.CLUSTER]: clusterEntities,
  [CloudProvider.AZURE]: azureEntities
}

const fieldToFieldNameMapping: Record<string, string> = {
  gcpProduct: 'Product',
  gcpProjectId: 'Project',
  gcpSKUDescription: 'SKUs',
  clusterName: 'Cluster Name',
  namespace: 'Namespace',
  workloadName: 'Workload',
  awsUsageAccountId: 'Account',
  awsServiceCode: 'Service',
  awsInstancetype: 'Instance Type',
  awsUsageType: 'Usage Type',
  workloadType: 'Workload Type',
  gcpSkuDescription: 'SKUs'
}

export function generateFilters(
  entityMap: Record<string, string>,
  cloudProvider: CloudProvider
): QlceViewFilterInput[] {
  const filters: any = []
  const relatedEntities = cloudProviderToEntityMapping[cloudProvider]
  relatedEntities.length &&
    relatedEntities.forEach(entity => {
      if (entityMap[entity]) {
        filters.push({
          field: {
            fieldId: entity,
            fieldName: fieldToFieldNameMapping[entity] || '',
            identifier: cloudProvider,
            identifierName: cloudProvider
          },
          operator: 'IN',
          type: 'VIEW_ID_CONDITION',
          values: [entityMap[entity]]
        })
      }
    })

  return filters
}

export function generateGroupBy(field: string, cloudProvider: CloudProvider): any {
  return field
    ? {
        fieldId: field,
        fieldName: fieldToFieldNameMapping[field] || '',
        identifier: cloudProvider,
        identifierName: cloudProvider
      }
    : DEFAULT_GROUP_BY
}

export function getCloudProviderFromFields(entityMap: Record<string, string | null>): CloudProvider {
  if (entityMap.gcpProjectId) {
    return CloudProvider.GCP
  }
  if (entityMap.awsUsageAccountId) {
    return CloudProvider.AWS
  }
  return CloudProvider.CLUSTER
}

export function getFiltersFromEnityMap(
  entityMap: Array<Record<string, string>>,
  cloudProvider: CloudProvider
): QlceViewFilterInput[] {
  const filters: any = []
  const relatedEntities = cloudProviderToEntityMapping[cloudProvider]

  relatedEntities.length &&
    relatedEntities.forEach(entity => {
      if (entityMap.length && entityMap[0][entity]) {
        const values: string[] = []
        entityMap.forEach(obj => {
          obj[entity] && values.push(obj[entity])
        })

        values.length &&
          filters.push({
            field: {
              fieldId: entity,
              fieldName: fieldToFieldNameMapping[entity] || '',
              identifier: cloudProvider,
              identifierName: cloudProvider
            },
            operator: 'IN',
            type: 'VIEW_ID_CONDITION',
            values: values
          })
      }
    })

  return filters
}
