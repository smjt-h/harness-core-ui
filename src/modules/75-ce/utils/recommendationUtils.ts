/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { IconName } from '@harness/uicore'
import type { IState } from '@ce/components/NodeRecommendation/constants'
import type { RecommendClusterRequest } from 'services/ce/recommenderService'

export const getProviderIcon = (provider: string): IconName => {
  const iconMapping: Record<string, IconName> = {
    google: 'gcp',
    azure: 'service-azure',
    amazon: 'service-aws'
  }

  return iconMapping[provider] || 'app-kubernetes'
}

export const addBufferToValue = (value: number, bufferPercentage: number, precision?: number): number =>
  +(((100 + bufferPercentage) / 100) * value).toFixed(precision || 2)

export const calculateSavingsPercentage = (savings: number, totalCost: number): string =>
  `(${Math.floor((savings / totalCost) * 100)}%)`

export const isResourceConsistent = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  buffer: number
): boolean => {
  const sumCpuWithBuffer = addBufferToValue(sumCpu, buffer)
  const sumMemWithBuffer = addBufferToValue(sumMemory, buffer)

  const isInconsistent =
    Math.round(sumCpuWithBuffer) < Math.round(maxCpu) || Math.round(sumMemWithBuffer) < Math.round(maxMemory)

  const anyZero =
    Math.round(sumCpuWithBuffer) === 0 ||
    Math.round(sumMemWithBuffer) === 0 ||
    Math.round(maxCpu) === 0 ||
    Math.round(maxMemory) === 0

  return !isInconsistent && !anyZero
}

export const calculateNodes = (
  sumCpu: number,
  sumMemory: number,
  maxCpu: number,
  maxMemory: number,
  minNodes: number
): { maximumNodes: number; minimumNodes: number } => {
  let minimumNodes = minNodes || 3

  let maximumNodes = Math.min(Math.floor(sumCpu / maxCpu), Math.floor(sumMemory / maxMemory))
  maximumNodes = Math.max(maximumNodes, 1)

  if (maximumNodes < minNodes) {
    minimumNodes = maximumNodes
  }

  return { maximumNodes, minimumNodes }
}

export const convertStateToRecommendClusterPayload = (
  state: IState,
  resourceRequirement: RecommendClusterRequest,
  buffer: number
): RecommendClusterRequest => {
  const sumCpuWithBuffer = addBufferToValue(state.sumCpu, buffer)
  const sumMemWithBuffer = addBufferToValue(state.sumMem, buffer)

  const { maximumNodes, minimumNodes } = calculateNodes(
    sumCpuWithBuffer,
    sumMemWithBuffer,
    state.maxCpu,
    +state.maxMemory,
    +state.minNodes
  )

  return {
    ...resourceRequirement,
    sumCpu: +sumCpuWithBuffer,
    sumMem: +sumMemWithBuffer,
    minCpu: +state.maxCpu,
    minMem: +state.maxMemory,
    maxNodes: maximumNodes,
    minNodes: minimumNodes,
    includeSeries: state.includeSeries,
    includeTypes: state.includeTypes,
    excludeSeries: state.excludeSeries,
    excludeTypes: state.excludeTypes
  }
}

export const addBufferToState = (state: IState, buffer: number): IState => ({
  ...state,
  sumCpu: addBufferToValue(state.sumCpu, buffer),
  sumMem: addBufferToValue(state.sumMem, buffer)
})

interface InstanceFamilies {
  includeSeries: string[]
  includeTypes: string[]
  excludeSeries: string[]
  excludeTypes: string[]
}

export const getInstanceFamiliesFromState = (state: IState): InstanceFamilies => ({
  includeSeries: state.includeSeries,
  includeTypes: state.includeTypes,
  excludeSeries: state.excludeSeries,
  excludeTypes: state.excludeTypes
})
