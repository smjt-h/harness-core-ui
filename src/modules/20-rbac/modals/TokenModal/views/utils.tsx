/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getReadableDateTime } from '@common/utils/dateUtils'

export const DATE_FORMAT = 'MM/DD/YYYY'

export const hasNoExpiryDate = (value?: number): boolean => {
  return getReadableDateTime(value, DATE_FORMAT) === '12/31/2099'
}

export const getSelectedExpiration = (value?: number): string => {
  if (!value) {
    return '30'
  } else if (hasNoExpiryDate(value)) {
    return '-1'
  }
  return 'custom'
}

export const getSelectedExpirationDate = (value: string): string => {
  switch (value) {
    case '30':
    case '90':
    case '180':
      return getReadableDateTime(new Date().setDate(new Date().getDate() + parseInt(value)), DATE_FORMAT)
    case '-1':
      return '12/31/2099'
    default:
      return ''
  }
}

export const getExpirationDate = (value?: number): string => {
  if (value) {
    return getReadableDateTime(value, DATE_FORMAT)
  } else {
    return getSelectedExpirationDate(getSelectedExpiration(value))
  }
}
