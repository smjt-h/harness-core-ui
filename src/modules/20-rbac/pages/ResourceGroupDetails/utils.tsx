/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import produce from 'immer'
import type { SelectOption } from '@wings-software/uicore'
import { RbacResourceGroupTypes } from '@rbac/constants/utils'
import type { ResourceType } from '@rbac/interfaces/ResourceType'
import { isDynamicResourceSelector, SelectionType } from '@rbac/utils/utils'
import type {
  ResourceSelector,
  ResponseResourceTypeDTO,
  ScopeSelector,
  ResourceGroupV2,
  ResourceSelectorV2,
  ResourceGroupV2Request,
  ResourceFilter
} from 'services/resourcegroups'
import { Scope } from '@common/interfaces/SecretsInterface'
import type { UseStringsReturn } from 'framework/strings'
import RbacFactory from '@rbac/factories/RbacFactory'
import { getScopeFromDTO } from '@common/components/EntityReference/EntityReference'

export enum SelectorScope {
  CURRENT = 'CURRENT',
  INCLUDE_CHILD_SCOPES = 'INCLUDE_CHILD_SCOPES',
  CUSTOM = 'CUSTOM'
}

enum ValidatorTypes {
  BY_RESOURCE_IDENTIFIER = 'BY_RESOURCE_IDENTIFIER',
  BY_RESOURCE_TYPE = 'BY_RESOURCE_TYPE',
  BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES = 'BY_RESOURCE_TYPE_INCLUDING_CHILD_SCOPES'
}

export const getSelectedResourcesMap = (
  resourceTypes: ResourceType[],
  resourceFilter?: ResourceFilter
): Map<ResourceType, string[] | string> => {
  const map: Map<ResourceType, string[] | string> = new Map()
  if (resourceFilter?.includeAllResources) {
    resourceTypes.forEach(resourceType => {
      map.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  } else {
    resourceFilter?.resources?.map(resourceSelector => {
      if (resourceSelector.identifiers?.length) {
        map.set(resourceSelector.resourceType as ResourceType, resourceSelector.identifiers)
      } else {
        map.set(resourceSelector.resourceType as ResourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
      }
    })
  }

  return map
}

export const validateResourceSelectors = (value: ResourceSelector[]): ResourceType[] => {
  return value
    .filter(val => {
      if (!isDynamicResourceSelector(val.type)) {
        if (val.identifiers?.length < 1) {
          return true
        }
      }
      return false
    })
    .map(val => val.resourceType)
}

export const getResourceSelectorsfromMap = (map: Map<ResourceType, string[] | string>): ResourceSelectorV2[] => {
  const resourceSelectors: ResourceSelectorV2[] = []
  map.forEach((value, key) => {
    if (isDynamicResourceSelector(value)) {
      resourceSelectors.push({
        resourceType: key
      })
    } else {
      resourceSelectors.push({
        resourceType: key,
        identifiers: value as string[]
      })
    }
  })
  return resourceSelectors
}

export const computeResourceMapOnChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  resourceType: ResourceType,
  isAdd: boolean,
  identifiers?: string[]
): void => {
  if (identifiers) {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, identifiers)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          if (Array.isArray(draft.get(resourceType))) {
            draft.set(
              resourceType,
              (draft.get(resourceType) as string[]).filter(el => !identifiers.includes(el))
            )
          }
          if (draft.get(resourceType)?.length === 0) {
            draft.delete(resourceType)
          }
        })
      )
    }
  } else {
    if (isAdd) {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      )
    } else {
      setSelectedResourceMap(
        produce(selectedResourcesMap, draft => {
          draft.delete(resourceType)
        })
      )
    }
  }
}

export const computeResourceMapOnMultiChange = (
  setSelectedResourceMap: (value: React.SetStateAction<Map<ResourceType, string | string[]>>) => void,
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  types: ResourceType[],
  isAdd: boolean
): void => {
  if (isAdd) {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
        })
      })
    )
  } else {
    setSelectedResourceMap(
      produce(selectedResourcesMap, draft => {
        types.forEach(resourceType => {
          draft.delete(resourceType)
        })
      })
    )
  }
}

export const getScopeLabelFromApi = (
  getString: UseStringsReturn['getString'],
  scope: Scope,
  resourceGroup: ResourceGroupV2
): string => {
  const selectorScope = getScopeType(resourceGroup)
  const dropDownItems = getScopeDropDownItems(scope, getString)
  const option = dropDownItems.filter(item => item.value === selectorScope)
  return option.length ? option[0].label : ''
}

export const getScopeDropDownItems = (scope: Scope, getString: UseStringsReturn['getString']): SelectOption[] => {
  switch (scope) {
    case Scope.PROJECT:
      return [{ label: getString('rbac.scopeItems.projectOnly'), value: SelectorScope.CURRENT }]
    case Scope.ORG:
      return [
        { label: getString('rbac.scopeItems.orgOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.orgAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
    case Scope.ACCOUNT:
    default:
      return [
        { label: getString('rbac.scopeItems.accountOnly'), value: SelectorScope.CURRENT },
        { label: getString('rbac.scopeItems.accountAll'), value: SelectorScope.INCLUDE_CHILD_SCOPES }
      ]
  }
}

export const getSelectionType = (resourceGroup?: ResourceGroupV2): SelectionType => {
  if (resourceGroup?.resourceFilter?.includeAllResources) {
    return SelectionType.ALL
  }
  return SelectionType.SPECIFIED
}
export const getScopeType = (resourceGroup?: ResourceGroupV2): SelectorScope => {
  if (resourceGroup?.includedScopes?.length) {
    for (const scope of resourceGroup?.includedScopes) {
      if (getScopeFromDTO(resourceGroup) === getScopeFromDTO(scope)) {
        return scope.filter === 'INCLUDING_CHILD_SCOPES' ? SelectorScope.INCLUDE_CHILD_SCOPES : SelectorScope.CURRENT
      } else {
        return SelectorScope.CUSTOM
      }
    }
  }

  return SelectorScope.CURRENT
}

export const getFormattedDataForApi = (
  data: ResourceGroupV2,
  selectionType: SelectionType,
  includedScopes: ScopeSelector[],
  resourceSelectors: ResourceSelectorV2[]
): ResourceGroupV2Request => {
  return {
    resourceGroup: {
      ...data,
      resourceFilter:
        selectionType === SelectionType.ALL
          ? {
              includeAllResources: true
            }
          : {
              resources: resourceSelectors,
              includeAllResources: false
            },
      includedScopes: includedScopes === [] ? getIncludedScopes(SelectorScope.CURRENT, data) : includedScopes
    }
  }
}

export const getFilteredResourceTypes = (
  data: ResponseResourceTypeDTO | null,
  selectedScope: SelectorScope
): ResourceType[] => {
  const resourceTypes = data?.data?.resourceTypes || []
  return resourceTypes?.reduce((acc: ResourceType[], value) => {
    if (
      value.name &&
      RbacFactory.isRegisteredResourceType(value.name as ResourceType) &&
      ((selectedScope === SelectorScope.CURRENT && value.validatorTypes?.includes(ValidatorTypes.BY_RESOURCE_TYPE)) ||
        selectedScope !== SelectorScope.CURRENT)
    ) {
      acc.push(value.name as ResourceType)
    }
    return acc
  }, [])
}

export const cleanUpResourcesMap = (
  resourceTypes: ResourceType[],
  selectedResourcesMap: Map<ResourceType, string | string[]>,
  selectionType: SelectionType
): Map<ResourceType, string | string[]> => {
  const types = [...resourceTypes]
  selectedResourcesMap.forEach((_value, key) => {
    if (!types.includes(key)) {
      selectedResourcesMap.delete(key)
    } else {
      types.splice(types.indexOf(key), 1)
      selectedResourcesMap.set(key, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    }
  })
  if (selectionType === SelectionType.ALL) {
    types.map(resourceType => {
      selectedResourcesMap.set(resourceType, RbacResourceGroupTypes.DYNAMIC_RESOURCE_SELECTOR)
    })
  }
  return selectedResourcesMap
}

export const getDefaultIncludedScope = (
  accountIdentifier: string,
  orgIdentifier?: string,
  projectIdentifier?: string,
  includedScopes?: ScopeSelector[]
): ScopeSelector[] => {
  if (includedScopes?.length) {
    return includedScopes
  }
  return [
    {
      accountIdentifier,
      orgIdentifier,
      projectIdentifier,
      filter: 'EXCLUDING_CHILD_SCOPES'
    }
  ]
}

export const getIncludedScopes = (type: SelectorScope, resourceGroup: ResourceGroupV2): ScopeSelector[] => {
  const { accountIdentifier, orgIdentifier, projectIdentifier } = resourceGroup
  switch (type) {
    case SelectorScope.CUSTOM:
      //Replace with custom scope once it is added
      return []
    case SelectorScope.INCLUDE_CHILD_SCOPES:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'INCLUDING_CHILD_SCOPES'
        }
      ]
    default:
      return [
        {
          accountIdentifier,
          orgIdentifier,
          projectIdentifier,
          filter: 'EXCLUDING_CHILD_SCOPES'
        }
      ]
  }
}
