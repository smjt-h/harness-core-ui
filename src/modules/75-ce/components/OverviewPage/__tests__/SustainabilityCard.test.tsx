/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import SustainabilityCard from '../SustainabilityCard'

const mockedCardDetails = {
  title: 'Sustainability card title',
  firstColValue: '1234',
  firstColText: 'Total',
  secondColValue: '1020',
  secondColText: 'Reduced'
}

describe('Sustainability Card test', () => {
  test('should show loader while fetching the data', () => {
    const { container } = render(
      <TestWrapper>
        <SustainabilityCard {...mockedCardDetails} fetching={true} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('should show card details after getting data', () => {
    const { container } = render(
      <TestWrapper>
        <SustainabilityCard {...mockedCardDetails} fetching={false} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
