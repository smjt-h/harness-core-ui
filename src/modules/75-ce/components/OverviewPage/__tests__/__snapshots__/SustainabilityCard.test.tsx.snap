// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sustainability Card test should show card details after getting data 1`] = `
<div>
  <div
    class="sustainabilityCard"
  >
    <div
      class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-flex-start"
    >
      <img
        src="test-file-stub"
        width="16"
      />
      <p
        class="StyledProps--font StyledProps--main StyledProps--font-variation-h6 StyledProps--color StyledProps--color-grey600"
      >
        Sustainability card title
      </p>
    </div>
    <div
      class="Layout--horizontal Layout--layout-spacing-huge StyledProps--main"
    >
      <div
        class="StyledProps--font StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-h4 StyledProps--color StyledProps--color-grey800"
        >
          <span>
            ce.common.emissionUnitHTML
          </span>
        </p>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-tiny StyledProps--color StyledProps--color-grey500"
        >
          Total
        </p>
      </div>
      <div
        class="StyledProps--font StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-h4 StyledProps--color StyledProps--color-grey800"
        >
          <span>
            ce.common.emissionUnitHTML
          </span>
        </p>
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-tiny StyledProps--color StyledProps--color-grey500"
        >
          Reduced
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Sustainability Card test should show loader while fetching the data 1`] = `
<div>
  <div
    class="sustainabilityCard"
  >
    <div
      class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-center"
    >
      <span
        class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-blue500"
        data-icon="spinner"
      >
        <svg
          fill="none"
          height="30"
          viewBox="0 0 36 36"
          width="30"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 2.084a15.915 15.915 0 010 31.831 15.915 15.915 0 010-31.83"
            stroke="currentColor"
            stroke-dasharray="75,100"
            stroke-linecap="round"
            stroke-width="2.8"
            style="animation: 1s linear infinite spinner-rotation; transform-origin: 50% 50%; transform-box: fill-box;"
          />
        </svg>
      </span>
    </div>
  </div>
</div>
`;
